### PR TITLE
APEXMALHAR-2139 

### PR DESCRIPTION
**Pull request by aayushia**
Emit only the keys for which values changed or were added in a given window.

### CODE REVIEW

**Comment by aayushia** [comment]
@devtagare please review

File: library/src/main/java/com/datatorrent/lib/testbench/CountAndLastTupleTestSink.java [hunk 1]
```diff
@@ -24,7 +24,7 @@
  * @displayName Count And Last Tuple Test Sink
  * @category Test Bench
  * @tags count, sink
- * @since 0.3.2
+ * @since 0.3.2
```
Comment (tweise) [inline library/src/main/java/com/datatorrent/lib/testbench/CountAndLastTupleTestSink.java]: Remove trailing whitespace.

File: library/src/main/java/com/datatorrent/lib/algo/UniqueCounter.java [hunk 2]
```diff
@@ -90,24 +91,30 @@ public void process(K tuple)
...
+        // Key present already, add the values
+        tuple.put(e.getKey(), e.getValue().toInteger());
+        oldMap.put(e.getKey(), e.getValue().toInteger() + oldMap.get(e.getKey()));
       }
-      tuple.put(e.getKey(), e.getValue().toInteger());
-    }
-    if (tuple != null) {
-      count.emit(tuple);
-    }
-    if (!cumulative) {
```
Comment (tweise) [inline library/src/main/java/com/datatorrent/lib/algo/UniqueCounter.java:108]: What is the purpose of cumulative after this change? When is map cleared?

**Comment by tweise** [comment]
@chinmaykolhatkar can you take a look at this PR?

**Comment by chinmaykolhatkar** [comment]
@tweise Sure.. Will do.

@aayushia Can you please explain what is the purpose of this change? I see there is a "cumulative" property of this operator which can assign downstream operator to identify the change.

**Comment by aayushia** [comment]
@chinmaykolhatkar The purpose is to emit only the changed keys in each window as compared to emitting the entire map in endWindow. I have made some changes to the code as per @tweise comment. Have to update the unit test. Please wait till I have made all the changes.

File: library/src/main/java/com/datatorrent/lib/testbench/CountAndLastTupleTestSink.java (hunk 1 above)
Comment (chinmaykolhatkar) [reply library/src/main/java/com/datatorrent/lib/testbench/CountAndLastTupleTestSink.java]: THis is not done yet.

File: library/src/main/java/com/datatorrent/lib/algo/UniqueCounter.java [hunk 3]
```diff
@@ -90,24 +91,30 @@ public void process(K tuple)
...
    */
   @Override
   public void endWindow()
   {
-    HashMap<K, Integer> tuple = null;
-    for (Map.Entry<K, MutableInt> e: map.entrySet()) {
+    HashMap<K, Integer> tuple = new HashMap<>();
+
+    for (Map.Entry<K, MutableInt> e : map.entrySet()) {
       if (tuple == null) {
```
Comment (chinmaykolhatkar) [inline library/src/main/java/com/datatorrent/lib/algo/UniqueCounter.java]: Not required. Tuple will never be null.

Comment (chinmaykolhatkar) [reply library/src/main/java/com/datatorrent/lib/algo/UniqueCounter.java]: you're just initializing it on line 99.

**Comment by chinmaykolhatkar** [comment]
@aayushia I am still not able to understand the reason for this PR. You said that you want to emit only the changed keys in each window. I think same can be achieved using the existing operator itself. If you set the "cumulative" to false which is the property of the operator, then the map will be cleared at every endWindow, making the "map" object a fresh one. In next endWindow, it'll have content only the ones which are received in the that window.

If you see the diff of the code, in endWindow, both if and else condition is adding the content of the "map" to the "tuple". So basically, you're adding everything of map to tuple and emitting to tuple.

Please correct if my understand is wrong here. But from the looks of it and based on your requirement, I see this change in PR is not required and your requirement can be achieved using existing operator.

**Comment by aayushia** [comment]
@chinmaykolhatkar I agree with what you said, the default test case checks with the "cumulative" property as false. However, irrespective, of it being true or false, it emits the entire map it got as input for that particular window. What I am tying to achieve, is emit only the keys which got modified or newly added in that window as compared to the previous windows (and not emit every key it received in that window). I have changed the code now, if you could kindly review it.

**Comment by chinmaykolhatkar** [comment]
@aayushia Thanks for the explanation. I'll review the PR.

File: library/src/main/java/com/datatorrent/lib/algo/UniqueCounter.java [hunk 4]
```diff
@@ -58,6 +56,7 @@
 public class UniqueCounter<K> extends BaseUniqueKeyCounter<K>
```
Comment (chinmaykolhatkar) [inline library/src/main/java/com/datatorrent/lib/algo/UniqueCounter.java:56]: javadoc for this method need to be updated for logic change done.

**Review by chinmaykolhatkar — CHANGES_REQUESTED** [review]
Please make the requested changes.
Then squash the commits and rebase with master branch.

**Review by chinmaykolhatkar — COMMENTED** [review]
Changes looks fine. Please squash the commits and rebase with master. I can merge the changes then.

**Comment by chinmaykolhatkar** [comment]
The checks passed for me locally. THe travis check has failed though. @aayushia After you squash commits and rebase with master please verify whether travis build also passes.

**Comment by aayushia** [comment]
@chinmaykolhatkar Made the change you mentioned. Squashed and rebased with master, as well. Can you please review and merge?

**Comment by chinmaykolhatkar** [comment]
@aayushia Your branch is not rebased with master, its very old. 
Please do "git fetch" on remote base repository and then only call "git rebase <remote>/master" on ur branch.
And then push the changes to ur remote repository.

This is important to avoid any unnecessary merge commits.

**Pull request merged**